### PR TITLE
Moved to JsonDocument

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
This greatly simplifies our JSON parser. CoreFx team just merged JsonDocument API and so we can stop using Utf8JsonReader. Yay!!!

@joshfree, @ahsonkhan, @bartonjs, @schaabs, @terrajobst 

@bartonjs, why did we decide to not have JsonElement.TryGet(string propertyName, out <type> value)? This would really simplify the code.

### CODE REVIEW

File: Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs [hunk 1]
```diff
@@ -243,8 +243,9 @@ public async Task Revisions()
 
                 int resultsReturned = 0;
                 SettingBatch batch = response.Result;
-                foreach (var value in batch)
+                for(int i=0; i<batch.Count; i++)
```
Comment (maririos) [inline Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs]: Why is this necessary?

Comment (maririos) [reply Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs]: I just saw your change deleting the enumerator. Never mind :)

File: Configuration/Azure.Configuration/ConfigurationSettingParser.cs [hunk 2]
```diff
@@ -43,196 +37,50 @@ public static bool TrySerialize(ConfigurationSetting setting, byte[] buffer, out
...
-            result = new List<ConfigurationSetting>();
-            consumed = 0;
-            var reader = new Utf8JsonReader(content, true, default);
-
-            JsonState state = JsonState.Other;
-            ConfigurationSetting value = default;
-            while (reader.Read())
-            {
-                switch (reader.TokenType)
+                for(int i=0; i<length; i++)
```
[the remark this answers is not in the document]
Comment (bartonjs) [reply Configuration/Azure.Configuration/ConfigurationSettingParser.cs]: > Why? Isn't each row in the database the same size, so you can skip from the beginning of the array (first row) to Nth row in O(1)?

Yes.  Every row is the same size.  A primitive value is one row.  An object's first row says how many rows all of the contained tokens occupy.  So when the array gets marked as "complex" finding the `[3]` element is

* look at 0, see how many rows to skip, jump
* look at 1, see how many rows to skip, jump
* look at 2, see how many rows to skip, jump
* return 3.

```C#

int i = 0;

foreach (JsonElement item in json.RootElement.GetProperty("items").EnumerateArray())
{
    settings[i] = ReadSetting(item);
    i++;
}

Debug.Assert(i == length);
```

[the remark this answers is not in the document]
Comment (KrzysztofCwalina) [reply Configuration/Azure.Configuration/ConfigurationSettingParser.cs]: Yeah, this is more general approach. But in many cases the array items will be the same size. I wonder if you could reserve a bit in the database (at the beginning of the array) and then write it base on whether all items are the same size or not, and then if they are, you could make the indexer O(1).

[the remark this answers is not in the document]
Comment (bartonjs) [reply Configuration/Azure.Configuration/ConfigurationSettingParser.cs]: All things are possible :smile:  Currently that's non-trivial since there's literally only one bit to spare in a start array record (and starting to use it adds masking operations to all reads), so either all records get bigger to hold that or arrays have to do something like "if the bit is set, and the index isn't 0, jump to the end record and see if we wrote down a heterogeneous size so we can skip the walk".

I don't have a good sense of how often the hybrid objects would have the same number of tokens (a single optional value or variable length nested array would invalidate the optimization) to know if it's worth the complexity.

**Comment by bartonjs** [comment]
> @bartonjs, why did we decide to not have JsonElement.TryGet(string propertyName, out value)? This would really simplify the code.

It's TryGetProperty.  If you're just looking to shorten the name that seems like an extension method that the JIT would be pretty good at inlining.

**Comment by KrzysztofCwalina** [comment]
@bartonjs, this is a mainline developer API. I think it would be good to make such that developers are delighted when they use it without the need for extension methods. 

cc: @terrajobst, @ahsonkhan
